Lib/importlib/_bootstrap_external.py

```diff
@@ -116,8 +116,8 @@ def _path_join(*path_parts):
             else:
                 root = new_root or root
                 path.append(tail)
-        path = [p.rstrip(path_separators) for p in path if p]
-        if len(path) == 1 and not path[0]:
+        path = [p for p in (p1.rstrip(path_separators) for p1 in path) if p]
+        if not path:
             # Avoid losing the root's trailing separator when joining with nothing
             return root + path_sep
         return root + path_sep.join(path)
```

Comment:
* `_path_join('//server/share', 'spam')` should return "//server/share/spam", not "//server/sharespam"
* `_path_join("C:/", "spam")` should return "C:/spam", not "C:spam"
* `_path_join("C:", "")` should return "C:", not "C:\\". 
* `_path_join("C:", "D:")` should return "D:", not "D:\\"

I don't like the change to make `_path_splitroot()` shift the separator into the path. It goes against what I expect given the function name, in terms of separating the root from the rest of the path. The `root` variable here actually has no root path, which is counterintuitive. 

As it stands, the following fixes the above case by shifting the separator back into `root`. The UNC + relative case has to be handled specially at the end.

```python
    if path and path[0].startswith(path_sep_tuple):
        root += path[0][0]
        path[0] = path[0][1:]
    tail = path_sep.join(c for p in path if (c := p.rstrip(path_separators)))
    # Fix up appending a relative path to a UNC root
    if (tail and root.startswith(path_sep_tuple) and
          not root.endswith(path_sep_tuple) and
          not tail.startswith(path_sep_tuple)):
        root += path_sep
    return root + tail
```